File: 05-async-cpp/sleeper.cpp
#include "sleeper.hpp"

#include <iostream>
#include <algorithm>

void Sleeper::worker(Sleeper* sleeper) {
    for(;;) {
        {
            const auto tp = std::chrono::system_clock::to_time_t(std::chrono::system_clock::now());
            std::scoped_lock lock(sleeper->sync_);
            while (!sleeper->queue_.empty() && sleeper->queue_.begin()->time_point_ < tp) {
                auto it = sleeper->queue_.begin();
                sleeper->pool_->run(std::move(it->func_));
                sleeper->queue_.erase(it);
            }
            if (sleeper->queue_.empty() && sleeper->shutdown_.load()) return;
        }
        using namespace std::chrono_literals;
        std::this_thread::sleep_for(100ms);
    }
}

Sleeper::Sleeper(ThreadPool* pool): pool_(pool) {
    std::thread t(worker, this);   
    this->worker_ = std::move(t); 
}

Sleeper::~Sleeper() {
    shutdown_.store(true);
    worker_.join();
}

void Sleeper::run(int ms, Closure func) {
    {
        std::scoped_lock lock(sync_);
        queue_.emplace(ms, std::move(func));
    }
}

std::atomic<int> Sleeper::TimePoint::counter;

Sleeper::TimePoint::TimePoint(int ms, Closure func) {
    time_point_ = std::chrono::system_clock::to_time_t(std::chrono::system_clock::now() 
                                                        + std::chrono::milliseconds(ms));
    func_ = std::move(func);
    id_ = counter.fetch_add(1);
}

bool Sleeper::TimePoint::operator<(const Sleeper::TimePoint& other) const {
    if (time_point_ == other.time_point_) {
        return id_ < other.id_;
    }
    return time_point_ < other.time_point_;
}

bool Sleeper::TimePoint::operator==(const Sleeper::TimePoint& other) const {
    return this == &other;
}

File: 06-async-rust/src/sleeper.rs
use crate::threadpool::PoolHandle;
use std::collections::BTreeSet;
use std::marker::PhantomData;
use std::sync::Arc;
use std::sync::atomic::Ordering::SeqCst;
use std::sync::{Mutex, atomic::AtomicBool};
use std::time::{Instant, Duration};
use std::cmp::{Eq, Ord, Ordering, PartialEq, PartialOrd};

pub trait Func = Fn() + Send + 'static;

struct Task {
    func: Box<dyn Func>,
    instant: Instant 
}

impl PartialEq for Task {
    fn eq(&self, other: &Self) -> bool {
        &*self.func as *const dyn Func == &*other.func as *const dyn Func
    }
}

impl Eq for Task {}

impl PartialOrd for Task {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.instant.cmp(&other.instant))
    }
}

impl Ord for Task {
    fn cmp(&self, other: &Self) -> Ordering {
        self.instant.cmp(&other.instant)
    }
}

struct Shared {
    queue: Mutex<BTreeSet<Task>>,
    terminated: AtomicBool,
    pool: PoolHandle<'static>
}

pub struct Sleeper<'a> {
    shared: Arc<Shared>,
    thread: Option<std::thread::JoinHandle<()>>,
    _lifetime: std::marker::PhantomData<&'a ()>
}

impl<'a> Sleeper<'a> {
    pub fn new(handle: PoolHandle<'a>) -> Self {
        let shared = Arc::new(Shared{
            queue: Mutex::new(BTreeSet::new()),
            terminated: AtomicBool::default(),
            pool: unsafe {std::mem::transmute::<PoolHandle<'a>, PoolHandle<'static>> (handle) }
        });
        let shared_c = shared.clone();
        Sleeper {
            shared,
            thread: Some(std::thread::spawn(move || Self::worker(shared_c))),
            _lifetime: PhantomData::default()
        }
    }
    fn worker(shared: Arc<Shared>) {
        loop {
            let mut lock = shared.queue.lock().unwrap();
            if shared.terminated.load(SeqCst) && lock.is_empty() { return; }
            let now = Instant::now();
            while let Some(task) = lock.first() {
                let task = match task.instant.cmp(&now) {
                    Ordering::Greater => break,
                    _ => lock.pop_first()
                };
                if let Some(task) = task {
                    shared.pool.send_boxed(task.func);
                }
            }
            drop(lock);
            std::thread::sleep(Duration::from_millis(5));
            std::hint::spin_loop();
        }
    }
    pub fn run(&self, delay: Duration, func: impl Func) {
        let mut g = self.shared.queue.lock().unwrap();
        g.insert(Task {
            func: Box::new(func),
            instant: Instant::now() + delay
        });
    }
}

impl<'a> Drop for Sleeper<'a> {
    fn drop(&mut self) {
        self.shared.terminated.store(true, SeqCst);
        self.thread.take().unwrap().join().ok();
    }
}

